crates/sui-graphql-rpc/src/types/transaction_block/tx_lookups.rs

```diff
@@ -0,0 +1,468 @@
+/// `network_total_transactions` - 1 of `hi_cp`.
+pub(crate) fn tx_bounds_query(lo_cp: u64, hi_cp: u64) -> RawQuery {
+    let lo = match lo_cp {
+        0 => query!("SELECT 0"),
+        _ => query!(format!(
+            r#"SELECT network_total_transactions
+            FROM checkpoints
+            WHERE sequence_number = {}"#,
+            lo_cp.saturating_sub(1)
+        )),
+    };
+
+    let hi = query!(format!(
+        r#"SELECT network_total_transactions - 1
+        FROM checkpoints
+        WHERE sequence_number = {}"#,
+        hi_cp
+    ));
+
+    query!(
+        "SELECT CAST(({}) AS BIGINT) AS lo, CAST(({}) AS BIGINT) AS hi",
+        lo,
+        hi
+    )
+}
+
+/// Determines the maximum value in an arbitrary number of Option<u64>.
+#[macro_export]
+macro_rules! max_option {
+    ($($x:expr),+ $(,)?) => {{
+        [$($x),*].iter()
+            .filter_map(|&x| x)
+            .max()
+    }};
+}
+
+/// Determines the minimum value in an arbitrary number of Option<u64>.
+#[macro_export]
+macro_rules! min_option {
+    ($($x:expr),+ $(,)?) => {{
+        [$($x),*].iter()
+            .filter_map(|&x| x)
+            .min()
+    }};
+}
+
+/// Constructs a `RawQuery` as a join over all relevant side tables, filtered on their own filter
+/// condition, plus optionally a sender, plus optionally tx/cp bounds.
+pub(crate) fn subqueries(filter: &TransactionBlockFilter, tx_bounds: TxBounds) -> Option<RawQuery> {
+    let sender = filter.sign_address;
+
+    let mut subqueries = vec![];
+
+    if let Some(f) = &filter.function {
+        subqueries.push(match f {
+            FqNameFilter::ByModule(filter) => match filter {
+                ModuleFilter::ByPackage(p) => (select_pkg(p, sender, tx_bounds), "tx_calls_pkg"),
+                ModuleFilter::ByModule(p, m) => {
+                    (select_mod(p, m.clone(), sender, tx_bounds), "tx_calls_mod")
+                }
+            },
+            FqNameFilter::ByFqName(p, m, n) => (
+                select_fun(p, m.clone(), n.clone(), sender, tx_bounds),
+                "tx_calls_fun",
+            ),
+        });
+    }
+    if let Some(kind) = &filter.kind {
+        subqueries.push((select_kind(*kind, tx_bounds), "tx_kinds"));
+    }
+    if let Some(recv) = &filter.recv_address {
+        subqueries.push((select_recipient(recv, sender, tx_bounds), "tx_recipients"));
+    }
+    if let Some(input) = &filter.input_object {
+        subqueries.push((select_input(input, sender, tx_bounds), "tx_input_objects"));
+    }
+    if let Some(changed) = &filter.changed_object {
+        subqueries.push((
+            select_changed(changed, sender, tx_bounds),
+            "tx_changed_objects",
+        ));
+    }
+    if let Some(sender) = &sender {
+        if filter.complex_filters() == 0 || filter.kind.is_some() {
+            subqueries.push((select_sender(sender, tx_bounds), "tx_senders"));
+        }
```

Comment:
I don't quite follow why we have the condition around `filter.kind.is_some()` here -- it seems to work when kind is programmable, but if kind is SystemTx, then we should omit the sender filter, because it's redundant.

The right place to deal with this logic is actually in `select_kind` which should devolve into a `select_sender` query if the kind provided is programmable and a sender has been provided.